### ISSUE TITLE
UserModExample - examples for config/ui

### ISSUE THREAD

**Issue by netmindz** [ewoudwijma/StarMod#50]
The UserModExample should show how you display a module in the UI that allows you to request values to be saved to the model/config

**Comment by netmindz** [comment]
We also need a nicer syntax for setting a label of the input than all this

```
ui->initPassword(parentVar, "pw", "", 32, false, [](JsonObject var, unsigned8 rowNr, unsigned8 funType) { switch (funType) { //varFun
    case f_UIFun:
      ui->setLabel(var, "Password");
      return true;
    default: return false;
  }});

```

**Comment by ewoudwijma** [comment]
Agree this is a lot of code overhead. The reason uiFun was added that storing this in the model json will create a very big model and it is only descriptive. So the ui now asks for this by triggering uiFun. 
so open for better ideas.
It is less worse if also other triggers are implemented here (changeFun etc), because it uses the same setup.

note that this whole structure is a way to implement class like functionality. Alternatively each initNumber etc is a (singleton) class with methods like fx.uiFun or fx.changeFun implemented . 
I did not do that because we have a lot of classes / instances in that case, but maybe reconsider this …

**Comment by ewowi** [comment]
Hi Will,, it took some time but the solution is in latest commit:

https://github.com/MoonModules/StarLight/commit/1388bdf302f7018f4c530f247f0af16e6e57148b

This is only for Effects.layers, but next commits will deal with all of Star

**Comment by netmindz** [comment]
Great. Good to hear there is some progress. I look forward to seeing the new user friendly syntax 

Do you have something in mind you are aiming for?

**Comment by ewowi** [comment]
What do you mean with aiming ? 
Om the longer term aiming for StarBase to be the Linux, macOS or windows for esp32 devices 😁

**Comment by netmindz** [comment]
We agree that this syntax isn't very nice

ui->initPassword(parentVar, "pw", "", 32, false, [](JsonObject var, unsigned8 rowNr, unsigned8 funType) { switch (funType) { //varFun
    case f_UIFun:
      ui->setLabel(var, "Password");
      return true;
    default: return false;
  }});

But do you know yet what the new one will be, once you have written it?

**Comment by ewowi** [comment]
ui->initPassword(parentVar, "password", "", 32, false, [](JsonObject var, unsigned8 rowNr, unsigned8 funType) { switch (funType) { //varFun
default: return false;
}});

the varFun will stay to catch events (until we find a more syntax friendly version. 
- it allows for code to be in one place
- It is needed for a minimal footprint of string related info (eg comments or dropdown values : using a string but not also needing long term memory to store that string 
- More optimizations in the pipeline

**Comment by ewowi** [comment]
And now this variable can be retrieved by using the parent id, which will be wifi in this case 
Eg findVar(“wifi”, “password”)
And the ui will show wifi password
And html id=“wifi.password”
So cleaner than just pw somewhere in the code

**Comment by ewowi** [comment]
And I also want to implement pubsub 

so you can say

Notify (“wifi”, “password”, onChange, ‘do something’…

**Comment by netmindz** [comment]
Yeah that would be really helpful to get the bidirectional changes between Star and HomeAssistant working in an efficient manner

**Comment by ewowi** [comment]
All setLabels except for two occasions where it is conditional completely gone now.
Code looks much cleaner, because the cryptic var names to preserve uniqueness gone now (uniqueness is now the combi of parent var name and var name (pid,id)
Both in StarBase and StarLight
Can we close the issue ?

**Comment by ewowi** [comment]
Big step in making the variable event functions more beautiful:

See https://github.com/ewowi/StarBase/commit/b492f6eddad0290403457565e6c37172d0d0ff1f

This is one of the steps to use the Variable class all over the place. It encapsulates the JsonObject var. This is the only variable of this class but it contains a ton of functions around it. Making the code which uses this more clean.

Next step will be moving getValue and setValue into this class...

**Comment by ewowi** [comment]
And by the way, StarBase and StarLight have next to the main branch now also a dev branch where all development will be done on, optionally branches from dev can be made for multi commit changes.

All merging from dev to main will be done by pull requests so the release changes will automatically be propagated with the commits made into main

And Star repos will also moving all definition stuff into cpp files.

So it takes a while but it will end up as a class book example of proper programming ;-)

**Comment by ewowi** [comment]
I am working on pub/sub implementation. This also means that the way a variable is defined can change as follows: 

Old style: 

```
    Variable currentVar = ui->initSelect(tableVar, "effect", (uint8_t)0, false, [this](EventArguments) { switch (eventType) {
      case onUI:
        variable.setComment("Effect to show");
        JsonArray options = variable.setOptions();
        ...
        return true;
      case onChange:
        ...
        return true;
      default: return false;
    }});
```

New style:

```
   Variable currentVar = ui->initSelect(tableVar, "effect", (uint8_t)0);

    currentVar.subscribe(onUI, [this](EventArguments) {
      variable.setComment("Effect to show");
      JsonArray options = variable.setOptions();
        ...
    });

    currentVar.subscribe(onChange, [this](EventArguments) {
        ...
    });
```

It looks cleaner, there is however one drawback:

Varevents: old: 123 x 16 = 1968 new: 6 x 28 = 168 (8 + 16 + 1)

This shows the memory impact of the old and new style. As the function argument is stored in a vector so it can be executed when the event happens. In the old way it takes 16 bytes and there are 123 functions stored (each variable had 0 or 1 function).
In the new way a variable can have more then one function and to make that work we now need 28 bytes per function as not only the function argument is stored but also the variable and event type (onUI, onChange, etc) as at the moment an event happens, it needs to check which functions have subscribed to that event.
In above example I implemented 6 subscribe events for testing but if old style is replaced by new style, the 123 functions will become more (one for each event) and take 28 bytes each. So estimate 250 * 28 = 7000 bytes (old was 1968 Bytes)

Also note this mechanism allows for more publish and subscribe between modules (e.g. the E131 module wants to know when the brightness changes ...), and each subscribe will take 28 bytes of memory more.

**Comment by netmindz** [comment]
I wouldn't worry too much about memory, I would say there are only going to be a relatively small number of subscriptions and potentially they could be for more than just a single primitive. 

E.g you would **NOT** have a subscription for every single dmx channel in multiple universes, just say Home Assistant subscribing to effect, palette and brightness so changes in the webui are shown in Home Assistant and vice versa.

A subscription could potentially be for multiple variables so rather than unchange handler for brightness, you would have one that received the name-value pairs then loop around the values and say it name==brightness ...
